File: Parallelism/parallel.c
//Josue Gomez Arteaga A01787212

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <sys/time.h>
#include <pthread.h>

int num_threads = 4;

//Return the current time in seconds, used to measure execution time
double now ()
{
    struct timeval tv;
    gettimeofday(&tv, NULL);
    return tv.tv_sec + tv.tv_usec / 1000000.0;
}


//Exercise 2 - Prime sum
//Return 1 if x is a prime number, 0 otherwise, using trial division
int is_prime (long x)
{
    if (x < 2)
    {
        return 0;
    }
    if (x == 2)
    {
        return 1;
    }
    long limit = (long) ceil(sqrt((double) x));
    for (long i = 2; i <= limit; i++)
    {
        if (x % i == 0)
        {
            return 0;
        }
    }
    return 1;
}

//Declare a structure to send data to the threads for exercise 2
typedef struct {
    int index;
    long first;
    long last;
    //A pointer to the slot where the thread leaves its partial sum
    unsigned long * partial_ptr;
} prime_data_t;

//This function adds all the prime numbers in one slice of the range
void * primeThread (void * data)
{
    prime_data_t * info = data;
    unsigned long sum = 0;
    for (long i = info->first; i <= info->last; i++)
    {
        if (is_prime(i))
        {
            sum += i;
        }
    }
    *(info->partial_ptr) = sum;
    pthread_exit(NULL);
}

//Sequential: test every number from 2 to n and add the primes
unsigned long prime_sum_seq (long n)
{
    unsigned long sum = 0;
    for (long i = 2; i <= n; i++)
    {
        if (is_prime(i))
        {
            sum += i;
        }
    }
    return sum;
}

//Parallel: each thread sums the primes of one slice of the range
unsigned long prime_sum_par (long n)
{
    long chunk = (n - 1) / num_threads;

    //Create the array of thread id's
    pthread_t * tid = malloc(num_threads * sizeof(pthread_t));
    //Create an array of data structures
    prime_data_t * data = malloc(num_threads * sizeof(prime_data_t));
    //Array to receive the partial sum of every thread
    unsigned long * partials = malloc(num_threads * sizeof(unsigned long));

    //Start all the threads
    for (int i = 0; i < num_threads; i++)
    {
        //Fill the struct for each of the threads
        data[i].index = i;
        data[i].first = 2 + (long) i * chunk;
        data[i].last = (i == num_threads - 1) ? n : data[i].first + chunk - 1;
        data[i].partial_ptr = &partials[i];
        //Create the thread, and pass the pointer to the structure
        pthread_create(&tid[i], NULL, primeThread, &data[i]);
    }

    //Wait for the threads to finish
    for (int i = 0; i < num_threads; i++)
    {
        pthread_join(tid[i], NULL);
    }

    //Add the partial sums returned by the threads
    unsigned long sum = 0;
    for (int i = 0; i < num_threads; i++)
    {
        sum += partials[i];
    }
    free(tid);
    free(data);
    free(partials);
    return sum;
}


//Exercise 3 - Compute pi
//Declare a structure to send data to the threads for exercise 3
typedef struct {
    int index;
    long first;
    long last;
    double width;
    //A pointer to the slot where the thread leaves its partial sum
    double * partial_ptr;
} pi_data_t;

//This function adds the rectangle heights for one slice
void * piThread (void * data)
{
    pi_data_t * info = data;
    double sum = 0.0;
    for (long i = info->first; i <= info->last; i++)
    {
        double mid = (i + 0.5) * info->width;
        double height = 4.0 / (1.0 + mid * mid);
        sum += height;
    }
    *(info->partial_ptr) = sum;
    pthread_exit(NULL);
}

//Sequential: rectangle method using a single loop over n rectangles
double compute_pi_seq (long n)
{
    double sum = 0.0;
    double width = 1.0 / n;
    for (long i = 0; i < n; i++)
    {
        double mid = (i + 0.5) * width;
        double height = 4.0 / (1.0 + mid * mid);
        sum += height;
    }
    return width * sum;
}

//Parallel: each thread sums the heights of one slice of the rectangles
double compute_pi_par (long n)
{
    double width = 1.0 / n;
    long chunk = n / num_threads;

    //Create the array of thread id's
    pthread_t * tid = malloc(num_threads * sizeof(pthread_t));
    //Create an array of data structures
    pi_data_t * data = malloc(num_threads * sizeof(pi_data_t));
    //Array to receive the partial sum of every thread
    double * partials = malloc(num_threads * sizeof(double));

    //Start all the threads
    for (int i = 0; i < num_threads; i++)
    {
        //Fill the struct for each of the threads
        data[i].index = i;
        data[i].first = (long) i * chunk;
        data[i].last = (i == num_threads - 1) ? n - 1 : data[i].first + chunk - 1;
        data[i].width = width;
        data[i].partial_ptr = &partials[i];
        //Create the thread, and pass the pointer to the structure
        pthread_create(&tid[i], NULL, piThread, &data[i]);
    }

    //Wait for the threads to finish
    for (int i = 0; i < num_threads; i++)
    {
        pthread_join(tid[i], NULL);
    }

    //Add the partial sums and multiply by the rectangle width
    double sum = 0.0;
    for (int i = 0; i < num_threads; i++)
    {
        sum += partials[i];
    }
    free(tid);
    free(data);
    free(partials);
    return width * sum;
}


//Main: time sequential vs parallel for each exercise and show speedup
//Usage: ./parallel [num_threads]
int main (int argc, char * argv[])
{
    if (argc > 1)
    {
        num_threads = atoi(argv[1]);
    }
    printf("Running on %d threads\n\n", num_threads);

    double t1, tp, t0;

    //Exercise 2 - Prime sum (n = 2000000)
    printf("Exercise 2\n");
    t0 = now();
    unsigned long s2 = prime_sum_seq(2000000);
    t1 = now() - t0;
    t0 = now();
    unsigned long p2 = prime_sum_par(2000000);
    tp = now() - t0;
    printf("Sequential: %.4f s\n", t1);
    printf("Parallel:   %.4f s\n", tp);
    printf("Speedup:    %.2fx\n", t1 / tp);
    printf("Results match: %s\n\n", (s2 == p2) ? "true" : "false");

    //Exercise 3 - Compute pi (n = 500000000)
    printf("Exercise 3\n");
    t0 = now();
    double s3 = compute_pi_seq(500000000L);
    t1 = now() - t0;
    t0 = now();
    double p3 = compute_pi_par(500000000L);
    tp = now() - t0;
    printf("Sequential: %.4f s  (pi = %.15f)\n", t1, s3);
    printf("Parallel:   %.4f s  (pi = %.15f)\n", tp, p3);
    printf("Speedup:    %.2fx\n\n", t1 / tp);

    //Finish this thread, and wait for any child threads
    pthread_exit(NULL);
}

File: Parallelism/parallel.exs
#Josue Gomez Arteaga A01787212

defmodule Sequential do

#exercise 2
#prime_sum/1: sums all prime numbers in 2..n using trial division
#filters the full range keeping only primes, then sums them
  def prime_sum(n) do
    2..n
    |> Enum.filter(&is_prime?/1)
    |> Enum.sum()
  end

#is_prime?/1: returns true if x is prime using trial division up to sqrt(x)
#only checks odd divisors starting at 3 to halve the number of divisions
  defp is_prime?(x) when x < 2, do: false
  defp is_prime?(2), do: true
  defp is_prime?(x) when rem(x, 2) == 0, do: false
  defp is_prime?(x) do
    limit = trunc(:math.sqrt(x))
    not Enum.any?(3..limit//2, fn i -> rem(x, i) == 0 end)
  end

#exercise 3
#compute_pi/1: approximates pi by the rectangle (midpoint) rule over n intervals
#for each rectangle of width 1/n, evaluates 4/(1+x^2) at the midpoint and sums
  def compute_pi(n) do
    width = 1.0 / n
    sum =
      0..(n - 1)
      |> Enum.reduce(0.0, fn i, acc ->
        mid = (i + 0.5) * width
        acc + 4.0 / (1.0 + mid * mid)
      end)
    width * sum
  end

end


defmodule Parallel do

#exercise 2
#prime_sum/1: parallel version - each Task sums the primes in one slice of 2..n
#partial sums are added together after all tasks complete
  def prime_sum(n) do
    num_tasks = System.schedulers_online()
    chunk = div(n - 1, num_tasks)
    tasks =
      for i <- 0..(num_tasks - 1) do
        first = 2 + i * chunk
        last  = if i == num_tasks - 1, do: n, else: first + chunk - 1
        Task.async(fn ->
          first..last
          |> Enum.filter(&is_prime?/1)
          |> Enum.sum()
        end)
      end
    tasks
    |> Enum.map(&Task.await(&1, :infinity))
    |> Enum.sum()
  end

#exercise 3
#compute_pi/1: parallel version - each Task sums the rectangle heights of one slice
#partial sums are combined and multiplied by the rectangle width
  def compute_pi(n) do
    num_tasks = System.schedulers_online()
    width = 1.0 / n
    chunk = div(n, num_tasks)
    tasks =
      for i <- 0..(num_tasks - 1) do
        first = i * chunk
        last  = if i == num_tasks - 1, do: n - 1, else: first + chunk - 1
        Task.async(fn ->
          first..last
          |> Enum.reduce(0.0, fn j, acc ->
            mid = (j + 0.5) * width
            acc + 4.0 / (1.0 + mid * mid)
          end)
        end)
      end
    sum =
      tasks
      |> Enum.map(&Task.await(&1, :infinity))
      |> Enum.sum()
    width * sum
  end

#is_prime?/1: returns true if x is prime using trial division up to sqrt(x)
#only checks odd divisors starting at 3 to halve the number of divisions
  defp is_prime?(x) when x < 2, do: false
  defp is_prime?(2), do: true
  defp is_prime?(x) when rem(x, 2) == 0, do: false
  defp is_prime?(x) do
    limit = trunc(:math.sqrt(x))
    not Enum.any?(3..limit//2, fn i -> rem(x, i) == 0 end)
  end

end


num_tasks = System.schedulers_online()
IO.puts("Running on #{num_tasks} schedulers\n")

#exercise 2 - prime sum (n = 1000000)
IO.puts("=== Exercise 2 - prime sum (n = 1000000) ===")
{t1, s2} = :timer.tc(fn -> Sequential.prime_sum(1_000_000) end)
{tp, p2} = :timer.tc(fn -> Parallel.prime_sum(1_000_000) end)
IO.puts("Sequential: #{Float.round(t1 / 1_000_000, 4)} s")
IO.puts("Parallel:   #{Float.round(tp / 1_000_000, 4)} s")
IO.puts("Speedup:    #{Float.round(t1 / tp, 2)}x")
IO.puts("Results match: #{s2 == p2}\n")

#exercise 3 - compute pi (n = 50000000)
IO.puts("=== Exercise 3 - compute pi (n = 50000000) ===")
{t1, s3} = :timer.tc(fn -> Sequential.compute_pi(50_000_000) end)
{tp, p3} = :timer.tc(fn -> Parallel.compute_pi(50_000_000) end)
IO.puts("Sequential: #{Float.round(t1 / 1_000_000, 4)} s")
IO.puts("Parallel:   #{Float.round(tp / 1_000_000, 4)} s")
IO.puts("Speedup:    #{Float.round(t1 / tp, 2)}x")
IO.puts("Results match: #{Float.round(s3, 10) == Float.round(p3, 10)}\n")
